Zend/zend_inheritance.c

```diff
@@ -1146,21 +1146,19 @@ static zend_always_inline inheritance_status do_inheritance_check_on_method_ex(
 		parent = proto;
 	}
 
-	if (!check_only && child->common.prototype != proto && child_zv) {
-		do {
-			if (child->common.scope != ce && child->type == ZEND_USER_FUNCTION) {
-				if (ce->ce_flags & ZEND_ACC_INTERFACE) {
-					/* Few parent interfaces contain the same method */
-					break;
-				} else {
-					/* op_array wasn't duplicated yet */
-					zend_function *new_function = zend_arena_alloc(&CG(arena), sizeof(zend_op_array));
-					memcpy(new_function, child, sizeof(zend_op_array));
-					Z_PTR_P(child_zv) = child = new_function;
-				}
-			}
-			child->common.prototype = proto;
-		} while (0);
+	if (!check_only && child->common.prototype != proto
+	 && (child_scope == ce || child_zv)
```

Comment:
actually, this (last line) condition should enable/disable copying but "prototype" should be always set.

Reply:
This still looks wrong. It's possible, that we need to set "prototype",  but don't have to SEPARATE method.

Reply:
I don't believe this is true. `zend_add_trait_method` calls `do_inheritance_check_on_method` with `child_zv` being `NULL` in two cases:

1. When the trait method is abstract. In that case, the method shouldn't actually be added to the class. Instead, we just verify the signature of the existing method matches the trait methods signature. Child here is the existing method of the class_even if it comes from a parent class. Your PR clones the method to allow the modification of `prototype`, but with this we may actually lose the existing `prototype` and set it to the abstract trait method instead, which would change visibility rules. I will try to demonstrate this in a test.
2. When overriding parent methods with methods inserted by traits. Child here is already copied, so there's no reason to copy it again. Modification to this method is already safe.